proposals/function-delegate.md

```diff
@@ -0,0 +1,195 @@
+# Function delegate
+
+* **Type**: Design proposal
+* **Author**: Marcin Moskała
+* **Contributors**: 
+* **Status**: Proposition
+* **Prototype**: Proposition
+
+## Feedback 
+
+Discussion of this proposal is held in [this issue](https://github.com/Kotlin/KEEP/issues/25).
+
+## Summary
+
+Support function delegates - delegae that can be used on to delegate function calls.
+
+## Description
+
+Allow to use function type as a function delegate:
+
+```
+interface MainView {
+    fun clearList()
+}
+
+class ListAdapter {
+    fun clear() {}
+}
+
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override fun clearList() by adapter::clear
+}
+```
+
+## Motivation / use cases
+
+Two main motivations behind this feature are:
+ * Simplify function calls passing that is opened for function signature changes. 
+ * Allow wrapper functions with state to add catching, synchronization and other functionalities that needs state. 
+Let's discusse them one after another. 
+
+### Function calls passing 
+
+When we are keeping strict architectures, like MVP or Clean Architecture, we often need to pass call from one class to another. For example, when Presenter is deciding that list needs to be changed it needs to call Activity function which is passing this call to adapter which is managing elements on list:
+
+```
+interface MainView {
+    fun deleteListElementAt(position: Int)
+}
+
+class ListAdapter {
+    fun deleteAt(position: Int) {}
+}
+
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override fun deleteListElementAt(position: Int) {
+        adapter.deleteAt(position)
+    }
+}
+```
+
+Simpler and more agile implementation would be if we could use function delegate instead:
+
+
+```
+class MainActivity : MainView {
+    val adapter = ListAdapter()
+    
+    override fun deleteListElementAt(position: Int) by adapter::deleteAt
+}
+```
+
+Similarly, when we are passing calls to change specific trait of layout, we need to pass function call to specific change on layout:
+
+```
+interface MainView {
+    fun setTitle(text: String)
+}
+
+class MainActivity: MainView {
+
+    override fun setTitle(text: String) {
+        titleView.setText(text)
+    }
+}
+```
+
+We would replace it with function delegate:
+
+```
+class MainActivity: MainView {
+
+    override fun setTitle(text: String) by titleView::setText
+}
+```
+
+Note, that when we are operaring on function type, we can use all functional features like currying or extension functions.
+
+### Function wrapper implementation
+
+Sometimes we need to add functionalities to function that needs to hold state. For example, when we need to add buffering to function. Let's say we have fibbonacci function:
+
+```
+fun fib(i: Int) = if (i <= 2) 1 else fib(i - 1) + fib(i - 1)
+```
+
+Without buffering it is very unefficient. Using function delegation we might add buffering this way:
+
+```
+fun <V, T> memorise(f: (V) -> T): (V) -> T {
```

Comment:
Did you mean "memoize"?

Reply:
You are right, `memoize` is a better name for this function